src/trio/testing/_exceptiongroup_util.py

```diff
@@ -0,0 +1,385 @@
+from __future__ import annotations
+
+import re
+import sys
+from types import TracebackType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Optional,
+    Pattern,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    final,
+    overload,
+)
+
+import _pytest
+import _pytest._code
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+E = TypeVar("E", bound=BaseException)
+E2 = TypeVar("E2", bound=BaseException)
+T = TypeVar("T")
+EE: TypeAlias = Union[Type[E], "ExpectedExceptionGroup[E]"]
+EEE: TypeAlias = Union[E, Type[E], "ExpectedExceptionGroup[E]"]
+
+
+# copy-pasting code between here and WIP pytest PR, which doesn't use future annotations
+# ruff: noqa: UP007
+
+
+# inherit from BaseExceptionGroup for the sake of typing the return type of raises.
+# Maybe it also works to do
+# `if TYPE_CHECKING: ExpectedExceptionGroup = BaseExceptionGroup`
+# though then we would probably need to support `raises(ExceptionGroup(...))`
+# class ExpectedExceptionGroup(BaseExceptionGroup[E]):
+@final
+class ExpectedExceptionGroup(Generic[E]):
+    # one might consider also accepting `ExpectedExceptionGroup(SyntaxError, ValueError)`
+    @overload
+    def __init__(self, exceptions: EEE[E], *args: EEE[E]):
+        ...
+
+    @overload
+    def __init__(self, exceptions: tuple[EEE[E], ...]):
+        ...
+
+    def __init__(self, exceptions: EEE[E] | tuple[EEE[E], ...], *args: EEE[E]):
+        if isinstance(exceptions, tuple):
+            if args:
+                raise ValueError(
+                    "All arguments must be exceptions if passing multiple positional arguments."
+                )
+            self.expected_exceptions = exceptions
+        else:
+            self.expected_exceptions = (exceptions, *args)
+        if not all(
+            isinstance(exc, (type, BaseException, ExpectedExceptionGroup))
+            for exc in self.expected_exceptions
+        ):
+            raise ValueError(
+                "All arguments must be exception instances, types, or ExpectedExceptionGroup."
+            )
+
+    # TODO: TypeGuard
+    def matches(
+        self,
+        exc_val: Optional[BaseException],
+    ) -> bool:
+        if exc_val is None:
+            return False
+        if not isinstance(exc_val, BaseExceptionGroup):
+            return False
+        if len(exc_val.exceptions) != len(self.expected_exceptions):
```

Comment:
Oh gah, I assumed this would be like `except*` where you're just checking if certain exceptions are in the final group. Is that more useful behavior? With stricter matching sometimes? (though... I can't think of a scenario when you would *want* that non-strict behavior).

Reply:
yeah I was confused that https://github.com/pytest-dev/pytest/pull/11424 works in that loose sense - it seems completely backwards to me and, for testing, almost never what you want.